### PR TITLE
[tt-train] Added mesh support

### PR DESCRIPTION
**Pull request by dmakoviichuk-tt**
### Problem description
In order to use multiple devices in future we need to use MeshDevice.
Added num_cache_entries to the MeshDevice.
### What's changed
MeshDevice 1x1 is used instead of Device for n150.

### Checklist
- [x] Post commit CI passes
- [x] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/core/ttnn_all_includes.hpp [hunk 1]
```diff
@@ -9,17 +9,18 @@
...
-#include <tt_metal/host_api.hpp>                                                                   // NOLINT
-#include <tt_metal/hostdevcommon/common_values.hpp>                                                // NOLINT
-#include <tt_metal/impl/device/device.hpp>                                                         // NOLINT
-#include <ttnn/core.hpp>                                                                           // NOLINT
-#include <ttnn/cpp/ttnn/operations/copy.hpp>                                                       // NOLINT
-#include <ttnn/cpp/ttnn/operations/core/core.hpp>                                                  // NOLINT
-#include <ttnn/device.hpp>                                                                         // NOLINT
+#include <common/bfloat16.hpp>                       // NOLINT
+#include <tests/tt_metal/test_utils/env_vars.hpp>    // NOLINT
+#include <tt_metal/common/base_types.hpp>            // NOLINT
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/ttnn_all_includes.hpp:14]: please correct format

File: tt-train/sources/ttml/core/mesh_device.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <memory>
+
+#include "ttnn_all_includes.hpp"
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/mesh_device.hpp:9]: use <>

File: tt-train/sources/ttml/core/mesh_device.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include "ttnn_all_includes.hpp"
+
+namespace ttml::core {
+
+MeshDevice::MeshDevice(int device_index) :
+    m_mesh_device(ttnn::distributed::api::open_mesh_device(
+        ttnn::distributed::MeshShape(1, 1),
+        DEFAULT_L1_SMALL_SIZE,
+        DEFAULT_TRACE_REGION_SIZE,
+        1,
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/mesh_device.cpp:16]: what this one means?

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 4]
```diff
@@ -360,7 +360,7 @@ int main(int argc, char **argv) {
         "{} Steps training time: {} s, cache entries: {}\n",
         max_steps,
         (double)duration / 1000000.,
-        device->num_program_cache_entries());
+        device->get_devices()[0]->num_program_cache_entries());
```
Comment (rfurko-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:363]: lets create another method `get_device(uint32_t index)`

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 5]
```diff
@@ -343,7 +343,7 @@ int main(int argc, char **argv) {
             fmt::print(
                 "Full step time {} ms, cache entries: {}\n",
                 (double)duration / 1000,
-                device->num_program_cache_entries());
+                device->get_devices()[0]->num_program_cache_entries());
```
Comment (rfurko-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:346]: use method proposed below

**Review by rfurko-tt — APPROVED** [review]
Overall looks good, please clean up and add helper methods.

**Review by cfjchu — APPROVED** [review]
looks great

**Pull request merged**